examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,635 @@
+print(f"Batch size: {batch_size}")
+print(f"Steps per epoch: {int(np.ceil(train_example_count / batch_size))}")
+train_dataset = get_dataset(os.path.join(tfrecords_dir, "train-*.tfrecord"), batch_size)
+valid_dataset = get_dataset(os.path.join(tfrecords_dir, "valid-*.tfrecord"), batch_size)
+# Create a learning rate scheduler callback.
+reduce_lr = keras.callbacks.ReduceLROnPlateau(
+    monitor="val_loss", factor=0.2, patience=3
+)
+# Create an early stopping callback.
+early_stopping = tf.keras.callbacks.EarlyStopping(
+    monitor="val_loss", patience=5, restore_best_weights=True
+)
+history = dual_encoder.fit(
+    train_dataset,
+    epochs=num_epochs,
+    validation_data=valid_dataset,
+    callbacks=[reduce_lr, early_stopping],
+)
+print("Training completed. Saving vision and text encoders...")
+vision_encoder.save("vision_encoder")
+text_encoder.save("text_encoder")
+print("Models are saved.")
+
+"""
+Plotting the training loss:
+"""
+
+plt.plot(history.history["loss"])
+plt.plot(history.history["val_loss"])
+plt.ylabel("Loss")
+plt.xlabel("Epoch")
+plt.legend(["train", "valid"], loc="upper right")
+plt.show()
+
+"""
+## Search for images using natural language queries
+
+We can then retrieve images corresponding to natural language queries via
+the following steps:
+
+1. Generate embeddings for the images by feeding them into the `vision_encoder`.
+2. Feed the natural language query to the `text_encoder` to generate a query embedding.
+3. Compute the similarity between the query embedding and the image embeddings
+in the index to retrieve the indices of the top matches.
+4. Look up the paths of the top matching images to display them.
+
+Note that, after training the `dual encoder`, the only the fine-tuned `vision_encoder`
+and `text_encoder` models will be used, while the `dual_encoder` model will be discarded.
+"""
+
+"""
+### Generate embeddings for the images
+
+We load the images and feed them into the `vision_encoder` to generate their embeddings.
+In large scale systems, this step is performed using a parallel data processing frameworks,
+such as [Apache Spark](https://spark.apache.org) or [Apache Beam](https://beam.apache.org).
+Generating the image embeddings will take a few minutes.
+"""
+print("Loading vision and text encoders...")
+vision_encoder = keras.models.load_model("vision_encoder")
+text_encoder = keras.models.load_model("text_encoder")
+print("Models are loaded.")
+
+
+def read_image(image_path):
+    raw_image = tf.io.read_file(image_path)
+    image_array = tf.image.decode_jpeg(raw_image, channels=3)
+    return tf.image.resize(image_array, (299, 299))
+
+
+print(f"Image embeddings generation started...")
+print(f"Generating embeddings for {len(image_paths)} images...")
+image_embeddings = vision_encoder.predict(
+    tf.data.Dataset.from_tensor_slices(image_paths).map(read_image).batch(512),
+    verbose=1,
+)
+print(f"Image embeddings generation completed.")
```

Comment:
Same here (and in a few more places in this file)

Reply:
Done